bin/uniffle-function.sh

```diff
@@ -0,0 +1,376 @@
+    IFS='@' read -ra brup <<< "${tmpa[$i]}"
+
+    option="${brup[0]}"
+    cmdtype="${brup[1]}"
+    giventext="${brup[2]}"
+
+    if [[ -n "${reqtype}" ]]; then
+      if [[ "${cmdtype}" != "${reqtype}" ]]; then
+        ((i=i+1))
+        continue
+      fi
+    fi
+
+    if [[ -z "${giventext}" ]]; then
+      giventext=${cmdtype}
+    fi
+
+    while read -r line; do
+      printf "%-${maxoptsize}s   %-s\n" "${option}" "${line}"
+      option=" "
+    done < <(echo "${giventext}"| fold -s -w ${foldsize})
+    ((i=i+1))
+  done
+}
+
+#---
+# uniffle_add_subcommand: Add a subcommand to the usage output.
+# @param1 subcommand
+# @param2 subcommandtype
+# @param3 subcommanddesc
+#---
+function uniffle_add_subcommand
+{
+  declare subcmd=$1
+  declare subtype=$2
+  declare text=$3
+
+  uniffle_add_array_param UNIFFLE_SUBCMD_USAGE_TYPES "${subtype}"
+
+  # done in this order so that sort works later
+  UNIFFLE_SUBCMD_USAGE[${UNIFFLE_SUBCMD_USAGE_COUNTER}]="${subcmd}@${subtype}@${text}"
+  ((UNIFFLE_SUBCMD_USAGE_COUNTER=UNIFFLE_SUBCMD_USAGE_COUNTER+1))
+}
+
+#---
+# uniffle_add_option: Add an option to the usage output.
+# @param1 subcommand
+# @param2 subcommanddesc
+#---
+function uniffle_add_option
+{
+  local option=$1
+  local text=$2
+
+  UNIFFLE_OPTION_USAGE[${UNIFFLE_OPTION_USAGE}]="${option}@${text}"
+  ((UNIFFLE_OPTION_USAGE_COUNTER=UNIFFLE_OPTION_USAGE_COUNTER+1))
+}
+
+#---
+# uniffle_java_setup: Configure/verify ${JAVA_HOME}
+# return may exit on failure conditions
+#---
+function uniffle_java_setup
+{
+  # Bail if we did not detect it
+  if [[ -z "${JAVA_HOME}" ]]; then
+    uniffle_error "ERROR: JAVA_HOME is not set and could not be found."
+    exit 1
+  fi
+
+  if [[ ! -d "${JAVA_HOME}" ]]; then
+    uniffle_error "ERROR: JAVA_HOME ${JAVA_HOME} does not exist."
+    exit 1
+  fi
+
+  JAVA="${JAVA_HOME}/bin/java"
+
+  if [[ ! -x "$JAVA" ]]; then
+    uniffle_error "ERROR: $JAVA is not executable."
+    exit 1
+  fi
+}
+
+#---
+# uniffle_java_exec: Execute the Java `class`, passing along any `options`.
+# @param1 command
+# @param2 class
+#---
+function uniffle_java_exec
+{
+  # run a java command.  this is used for
+  # non-daemons
+
+  local command=$1
+  local class=$2
+  shift 2
+
+  export CLASSPATH
+  exec "${JAVA}" "-Dproc_${command}" "${class}" "$@"
+}
+
+#---
+# uniffle_validate_classname: Verify that a shell command was passed a valid class name.
+# @param1 command
+# @return 0 = success; 1 = failure w/user message;
+#---
+function uniffle_validate_classname
+{
+  local class=$1
+  shift 1
+
+  if [[ ! ${class} =~ \. ]]; then
+    # assuming the arg is typo of command if it does not conatain ".".
+    # class belonging to no package is not allowed as a result.
+    uniffle_error "ERROR: ${class} is not COMMAND nor fully qualified CLASSNAME."
+    return 1
+  fi
+  return 0
+}
+
+function uniffle_generic_java_subcmd_handler
+{
+  uniffle_java_exec "${UNIFFLE_SUBCMD}" "${UNIFFLE_CLASSNAME}" "${UNIFFLE_SUBCMD_ARGS[@]}"
+}
```

Comment:
Add a new line.  There is a warning.

Reply:
Thanks for helping to review the code, I will fix it.